src/CPUActivityAgent.cpp

```diff
@@ -0,0 +1,532 @@
+        double f_uncore_range = in_policy[M_POLICY_UNCORE_FREQ_MAX] - in_policy[M_POLICY_UNCORE_FREQ_EFFICIENT];
+
+        //m_resolved_f_uncore_efficient = 0;
+        for (int domain_idx = 0; domain_idx < M_NUM_PACKAGE; ++domain_idx) {
+            double uncore_freq = (double) m_uncore_freq_status.at(domain_idx).value;
+
+            /////////////////////////////////////////////
+            // L3 Total External Bandwidth Measurement //
+            /////////////////////////////////////////////
+            auto qm_max_itr = m_qm_max_rate.lower_bound(uncore_freq);
+            if(qm_max_itr != m_qm_max_rate.begin()) {
+                qm_max_itr = std::prev(qm_max_itr, 1);
+            }
+
+            double scalability_uncore = 1.0;
+
+            // Handle divided by zero, either numerator or
+            // denominator being NAN, and the un-characterized case
+            if (!std::isnan(m_qm_rate.at(domain_idx).value) &&
+                !std::isnan(qm_max_itr->second) &&
+                qm_max_itr->second != 0 &&
+                m_qm_max_rate.size() != 0) {
+                scalability_uncore = (double) m_qm_rate.at(domain_idx).value /
+                                         qm_max_itr->second;
+            }
+
+            // L3 usage, Network Traffic, HBM, and PCIE (GPUs) all use the uncore.
+            // Eventually all these components should be considered when scaling
+            // the uncore frequency in the efficient - performant range.
+            // A more robust/future proof solution may be to directly query uncore
+            // counters that indicate utilization (when/if available).
+            // For now only L3 bandwith metric is used.
+            double uncore_req = m_resolved_f_uncore_efficient + f_uncore_range * scalability_uncore;
+
+            //Clip uncore request within policy limits
+            uncore_req = std::max(m_resolved_f_uncore_efficient, uncore_req);
+            uncore_req = std::min(m_resolved_f_uncore_max, uncore_req);
+            uncore_freq_request.push_back(uncore_req);
+        }
+
+        // Per core freq
+        std::vector<double> core_freq_request;
+        m_resolved_f_core_efficient = in_policy[M_POLICY_CPU_FREQ_EFFICIENT];
+        m_resolved_f_core_max = in_policy[M_POLICY_CPU_FREQ_MAX];
+        double f_core_range = in_policy[M_POLICY_CPU_FREQ_MAX] - in_policy[M_POLICY_CPU_FREQ_EFFICIENT];
+
+        for (int domain_idx = 0; domain_idx < M_NUM_CORE; ++domain_idx) {
+            //////////////////////////////////
+            // Core Scalability Measurement //
+            //////////////////////////////////
+            double scalability = (double) m_core_scal.at(domain_idx).value;
+            if (std::isnan(scalability)) {
+                scalability = 1.0;
+            }
+
+            double core_req = m_resolved_f_core_efficient + f_core_range * scalability;
+            //Clip core request within policy limits
```

Comment:
Debug warn?

Reply:
Moved to a report header following the discussion on the (similar) GPU compute activity agent clipping tracking [here](https://github.com/geopm/geopm/pull/2045#discussion_r871859791) as part of #2045